chan.go

```diff
@@ -128,92 +157,103 @@ func (r *chanResponse) RawNext() (interface{}, error) {
 
 		return nil, io.EOF
 	case <-ctx.Done():
-		close(r.done)
 		return nil, ctx.Err()
 	}
 
 }
 
-type chanResponseEmitter struct {
-	ch   chan<- interface{}
-	wait chan struct{}
-	done <-chan struct{}
-
-	length *uint64
-	err    **cmdkit.Error
-
-	emitted bool
-}
+type chanResponseEmitter chanResponse
 
 func (re *chanResponseEmitter) Emit(v interface{}) error {
+	// Initially this library allowed commands to return errors by sending an
+	// error value along a stream. We removed that in favour of CloseWithError,
+	// so we want to make sure we catch situations where some code still uses the
+	// old error emitting semantics.
+	// Also errors may occur both as pointers and as plain values, so we need to
+	// check both.
+	if e, ok := v.(cmdkit.Error); ok {
+		v = &e
+	}
+	if e, ok := v.(*cmdkit.Error); ok {
+		log.Errorf("unexpected error value emitted: %s at\n%s", e.Error(), debug.Stack())
+	}
+
+	// channel emission iteration
+	// TODO maybe remove this and use EmitChan instead of calling Emit directly?
 	if ch, ok := v.(chan interface{}); ok {
 		v = (<-chan interface{})(ch)
 	}
-
 	if ch, isChan := v.(<-chan interface{}); isChan {
-		for v = range ch {
-			err := re.Emit(v)
-			if err != nil {
-				return err
-			}
-		}
-		return nil
+		return EmitChan(re, ch)
 	}
 
-	re.emitted = true
-	if re.wait != nil {
+	// unblock Length(), Error() and Head()
+	re.waitOnce.Do(func() {
 		close(re.wait)
-		re.wait = nil
-	}
+	})
+
+	re.l.Lock()
+	defer re.l.Unlock()
 
 	if re.ch == nil {
 		return fmt.Errorf("emitter closed")
 	}
 
 	if _, ok := v.(Single); ok {
-		defer re.Close()
+		defer re.closeWithError(io.EOF)
 	}
 
+	ctx := re.req.Context
+
 	select {
 	case re.ch <- v:
 		return nil
-	case <-re.done:
-		return context.Canceled
-	}
-}
-
-func (re *chanResponseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
-	err := re.Emit(&cmdkit.Error{Message: fmt.Sprint(v), Code: errType})
-	if err != nil {
-		panic(err)
+	case <-ctx.Done():
+		return ctx.Err()
 	}
 }
 
 func (re *chanResponseEmitter) SetLength(l uint64) {
+	re.l.Lock()
+	defer re.l.Unlock()
+
 	// don't change value after emitting
 	if re.emitted {
 		return
 	}
 
-	*re.length = l
+	re.length = l
 }
 
-func (re *chanResponseEmitter) Head() Head {
-	<-re.wait
+func (re *chanResponseEmitter) CloseWithError(err error) error {
+	re.l.Lock()
+	defer re.l.Unlock()
 
-	return Head{
-		Len: *re.length,
-		Err: *re.err,
-	}
+	return re.closeWithError(err)
 }
 
-func (re *chanResponseEmitter) Close() error {
-	if re.ch == nil {
-		return nil
+func (re *chanResponseEmitter) closeWithError(err error) error {
+	if e, ok := err.(cmdkit.Error); ok {
+		err = &e
 	}
 
-	close(re.ch)
-	re.ch = nil
+	re.closeOnce.Do(func() {
+		re.err = err
+		close(re.ch)
+		close(re.closed)
+	})
+
+	// unblock Length(), Error() and Head()
+	re.waitOnce.Do(func() {
```

Comment:
Same here. If we make sure to always close this under a lock (need to modify `Emit` a bit), we can just use select to see if it's still open.

Reply:
Again, I don't think doing this with channel operations is more readable or efficient.

Reply:
For me, every synchronization primitive is something else I have to keep track of. It's easier to keep track of one big lock (and possibly some booleans) than a bunch of `sync.Once`s.

Reply:
Unfortunately "one big lock" doesn't work in our case (because we need to `Emit` and `Next` at the same time), but I'm down to two and removed all `Once`s.